### PR TITLE
paramiko connection now uses the correct source of data

### LINKED ISSUE #78509: Jinja expressions are not evaluated in credential variables if paramiko plugin is used
### Summary

I use paramiko ssh plugin. I try to set `ansible_ssh_user`, `ansible_ssh_pass`, etc. from jinja expression. This has worked correctly in version 2.10, jinja expressions were evaluated. But in ansible-core 2.13 jinja expressions are not evaluated:

`ESTABLISH PARAMIKO SSH CONNECTION FOR USER: {{ some_variable | default('cirros') }} on PORT 22 TO 10.xx.xx.xx`

### Issue Type

Bug Report

### Component Name

ansible-playbook,paramiko

### Ansible Version

```console
$ ansible --version
ansible [core 2.13.1]
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/zkrakko/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/zkrakko/ansible-venv/lib/python3.8/site-packages/ansible
  ansible collection location = /home/zkrakko/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/zkrakko/ansible-venv/bin/ansible
  python version = 3.8.10 (default, Jun 22 2022, 20:18:18) [GCC 9.4.0]
  jinja version = 3.1.2
  libyaml = True
```

### Configuration

```console
# if using a version older than ansible-core 2.12 you should omit the '-t all'
$ ansible-config dump --only-changed -t all
```

### OS / Environment

Linux Mint 20.3

### Steps to Reproduce

test.yaml file:
```yaml (paste below)
---
- hosts: server
  gather_facts: no
  vars:
    ansible_ssh_user: "{{ some_variable | default('cirros') }}"
    ansible_ss_pass: gocubsgo
  tasks:
    - raw: echo "{{ ansible_ssh_user }}"
```
hosts file:
```
[server]
10.xx.xx.xx
```
command:
`ansible-playbook -vvv -i hosts -c paramiko test.yaml`

### Expected Results

I expected that `cirros` username is used for connection (like in ansible-core 2.10):

```
PLAYBOOK: test.yaml ****************************************************************
1 plays in test.yaml

PLAY [server] **********************************************************************
META: ran handlers

TASK [raw] *************************************************************************
task path: /home/zkrakko/ansible-venv/test.yaml:8
<10.xx.xx.xx> ESTABLISH PARAMIKO SSH CONNECTION FOR USER: cirros on PORT 22 TO 10.xx.xx.xx
```

### Actual Results

```console
Instead, the jinja expression in username was not evaluated:

PLAYBOOK: test.yaml ****************************************************************
1 plays in test.yaml

PLAY [server] **********************************************************************
META: ran handlers

TASK [raw] *************************************************************************
task path: /home/zkrakko/ansible-venv/test.yaml:8
<10.xx.xx.xx> ESTABLISH PARAMIKO SSH CONNECTION FOR USER: {{ some_variable | default('cirros') }} on PORT 22 TO 10.xx.xx.xx
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### LINKED ISSUE #79638: inventory variable replace failed
### Summary

Variables in inventory after version 2.13. x can not be replaced with actual values when execute task,  because of this MR: https://github.com/ansible/ansible/pull/76590

config：
primaryIp: 'aa.aa.aa.aa'
standbyIp: 'bb.bb.bb.bb'
user_name: xx

inventory:
[xx]
primary ansible_ssh_user={{user_name}} ansible_ssh_host={{primaryIp}}
standby ansible_ssh_user={{user_name}} ansible_ssh_host={{standbyIp}}

expected result:
image
Actual Results:
image

### Issue Type

Bug Report

### Component Name

lib/ansible/executor/task_executor.py

### Ansible Version

```console
$ ansible --version
ansible [core 2.13.6]
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/xxx/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /opt/ansible/lib64/python3.9/site-packages/ansible
  ansible collection location = /home/xxx/.ansible/collections:/usr/share/ansible/collections
  executable location = /opt/ansible/bin/ansible
  python version = 3.9.9 (main, Sep 21 2022, 09:00:34) [GCC 10.3.1]
  jinja version = 3.0.3
  libyaml = True
```

### Configuration

```console
# if using a version older than ansible-core 2.12 you should omit the '-t all'
$ ansible-config dump --only-changed -t all
CACHE_PLUGIN(/etc/ansible/ansible.cfg) = memory
COMMAND_WARNINGS(/etc/ansible/ansible.cfg) = False
DEFAULT_FORKS(/etc/ansible/ansible.cfg) = 20
DEFAULT_GATHERING(/etc/ansible/ansible.cfg) = smart
DEFAULT_HOST_LIST(/etc/ansible/ansible.cfg) = ['/etc/ansible/hosts']
DEFAULT_MANAGED_STR(/etc/ansible/ansible.cfg) = Ansible managed: {file} modified on %Y-%m-%d %H:%M:%S by {uid} on {host}
DEFAULT_NO_TARGET_SYSLOG(/etc/ansible/ansible.cfg) = True
DEFAULT_POLL_INTERVAL(/etc/ansible/ansible.cfg) = 15
DEFAULT_TIMEOUT(/etc/ansible/ansible.cfg) = 60
DEFAULT_TRANSPORT(/etc/ansible/ansible.cfg) = paramiko
DEPRECATION_WARNINGS(/etc/ansible/ansible.cfg) = False
HOST_KEY_CHECKING(/etc/ansible/ansible.cfg) = False
INTERPRETER_PYTHON(/etc/ansible/ansible.cfg) = auto_legacy_silent
PARAMIKO_HOST_KEY_AUTO_ADD(/etc/ansible/ansible.cfg) = True
PARAMIKO_LOOK_FOR_KEYS(/etc/ansible/ansible.cfg) = False
RETRY_FILES_ENABLED(/etc/ansible/ansible.cfg) = True
SYSTEM_WARNINGS(/etc/ansible/ansible.cfg) = False

CONNECTION:
==========

paramiko_ssh:
____________
host_key_auto_add(/etc/ansible/ansible.cfg) = True
host_key_checking(/etc/ansible/ansible.cfg) = False
look_for_keys(/etc/ansible/ansible.cfg) = False
ssh_args(/etc/ansible/ansible.cfg) = -C -o ControlMaster=auto -o ControlPersist=1800s

ssh:
___
host_key_checking(/etc/ansible/ansible.cfg) = False
pipelining(/etc/ansible/ansible.cfg) = True
scp_if_ssh(/etc/ansible/ansible.cfg) = smart
ssh_args(/etc/ansible/ansible.cfg) = -C -o ControlMaster=auto -o ControlPersist=1800s
timeout(/etc/ansible/ansible.cfg) = 60

SHELL:
=====

sh:
__
remote_tmp(/etc/ansible/ansible.cfg) = ~/.ansible/tmp
```

### OS / Environment

EulerOS

### Steps to Reproduce

```yaml (paste below)
any
```
config：
primaryIp: 'aa.aa.aa.aa'
standbyIp: 'bb.bb.bb.bb'
user_name: xx

inventory:
[xx]
primary ansible_ssh_user={{user_name}} ansible_ssh_host={{primaryIp}}
standby ansible_ssh_user={{user_name}} ansible_ssh_host={{standbyIp}}

### Expected Results

ssh normal:
image

### Actual Results

```console
version after 2.13.x:

ansible_ssh_host can not be replaced with actual IP，ssh {{primaryIp}} or {{standbyip}}, so failed with:

<{{standbyIp}}> ESTABLISH PARAMIKO SSH CONNECTION FOR USER: {{user_name}} on PORT 22 TO {{standbyIp}}
fatal: [primary]: UNREACHABLE! => {
    "changed": false,
    "msg": "[Errno -2] Name or service not known",
    "unreachable": true
}
Traceback (most recent call last):
  File "/opt/ansible/lib64/python3.9/site-packages/ansible/plugins/connection/paramiko_ssh.py", line 345, in _connect_uncached
    ssh.connect(
  File "/opt/ansible/lib64/python3.9/site-packages/paramiko/client.py", line 340, in connect
    to_try = list(self._families_and_addresses(hostname, port))
  File "/opt/ansible/lib64/python3.9/site-packages/paramiko/client.py", line 203, in _families_and_addresses
    addrinfos = socket.getaddrinfo(
  File "/usr/lib64/python3.9/socket.py", line 954, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -2] Name or service not known
fatal: [standby]: UNREACHABLE! => {
    "changed": false,
    "msg": "[Errno -2] Name or service not known",
    "unreachable": true
}
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by bcoca**
##### SUMMARY
fixes #79638
fixes #78509

Also added missing 'timeout' parameter with existing options + new paramiko specific ones

related #76156 (would have originally fixed this)

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

connection/paramiko_ssh

### CODE REVIEW

File: changelogs/fragments/paramiko_config.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - paramiko connection was still using outdata playcontext, this should bring it up to date to use the 'correct' data for each task/loop.
```
Comment (s-hertel) [inline changelogs/fragments/paramiko_config.yml]:
```suggestion
  - paramiko connection was still using outdated playcontext, this should bring it up to date to use the 'correct' data for each task/loop.
```

**Review by s-hertel — APPROVED** [review]
You could add a test case for templating ansible_host in test/integration/targets/connection_paramiko_ssh/test_connection.inventory.

File: lib/ansible/plugins/connection/paramiko_ssh.py [hunk 2]
```diff
@@ -314,8 +359,15 @@ class Connection(ConnectionBase):
     transport = 'paramiko'
     _log_channel = None
 
+    def get_option(self, option, hostvars=None):
+        try:
+            r = super().get_option(option, hostvars=hostvars)
+        except Exception:
+            r = getattr(self._play_context, option)
```
Comment (jborean93) [inline lib/ansible/plugins/connection/paramiko_ssh.py]: Is there still somethign that is relying on a play context attribute? Wouldn't it best to remove that?

File: lib/ansible/plugins/connection/paramiko_ssh.py [hunk 3]
```diff
@@ -420,7 +472,7 @@ def _connect_uncached(self):
 
         ssh.set_missing_host_key_policy(MyAddPolicy(self._new_stdin, self))
 
-        conn_password = self.get_option('password') or self._play_context.password
+        conn_password = self.get_option('password') or self.get_option('password')
```
Comment (jborean93) [inline lib/ansible/plugins/connection/paramiko_ssh.py]: This is doubled up here

```suggestion
        conn_password = self.get_option('password')
```

File: lib/ansible/plugins/connection/paramiko_ssh.py [hunk 4]
```diff
@@ -429,25 +481,25 @@ def _connect_uncached(self):
 
         try:
             key_filename = None
-            if self._play_context.private_key_file:
-                key_filename = os.path.expanduser(self._play_context.private_key_file)
+            if self.get_option('private_key_file'):
```
Comment (jborean93) [inline lib/ansible/plugins/connection/paramiko_ssh.py:498]: `private_key_file` is not defined in the `DOCUMENTATION` block

File: lib/ansible/plugins/connection/paramiko_ssh.py (hunk 2 above)
Comment (bcoca) [reply lib/ansible/plugins/connection/paramiko_ssh.py]: using it as a fallback, JIC someone tries to use plugin on older ansible version

File: lib/ansible/plugins/connection/paramiko_ssh.py (hunk 4 above)
Comment (bcoca) [reply lib/ansible/plugins/connection/paramiko_ssh.py:498]: fixed

File: lib/ansible/plugins/connection/paramiko_ssh.py (hunk 2 above)
Comment (jborean93) [reply lib/ansible/plugins/connection/paramiko_ssh.py]: If they are copying the code then they are somewhat on the hook for any problems that occur. This should be tied to this Ansible version so I would be reluctant to have this fallback which might hide even more play context fallback problems we won't hit until later.

Comment (bcoca) [reply lib/ansible/plugins/connection/paramiko_ssh.py]: most plugins should not be 'hard' tied to the Ansible version, we've done this same migration several times fro other plugins and always left a fallback. In the case of paramiko we also have to account for the 'networking flow' which relies heavily on play_context and only recently also started migrating to the config system.

Comment (jborean93) [reply lib/ansible/plugins/connection/paramiko_ssh.py]: That's the problem, if this is "paramiko connection now uses the correct source of data" then it shouldn't be falling back on the incorrect source of data. We ship this with Ansible and I don't see us at any point putting it in a collection. Regardless the option model shouldn't be tied to an Ansible version but rather that's in the plugins documentation block. I just see having this fallback as a code smell, we now have this code that people don't know if it's needed or not or whether it's left over from something in the past. By removing the fallback we might encounter some issues in the future but at least we can see that the plugin is no longer dependent on play context altogether.

File: lib/ansible/plugins/connection/paramiko_ssh.py [hunk 5]
```diff
@@ -314,8 +380,11 @@ class Connection(ConnectionBase):
     transport = 'paramiko'
     _log_channel = None
 
+    def get_option(self, option, hostvars=None):
+        return super().get_option(option, hostvars=hostvars)
+
```
Comment (jborean93) [inline lib/ansible/plugins/connection/paramiko_ssh.py]: Is this needed anymore now that it's just calling super

```suggestion
```

**Review by jborean93 — APPROVED** [review]

**Pull request merged**